File: Python/LC669_Trim_BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    # Definition of BST should require no duplicated tree values.
    # Runtime: 52 ms, faster than 94.01% of Python3 online submissions for Trim a Binary Search Tree.
    # Memory Usage: 17.9 MB, less than 7.14% of Python3 online submissions for Trim a Binary Search Tree. (The solution use the same memory)
    def trimBST(self, root: TreeNode, L: int, R: int) -> TreeNode:
        # def trimBSTHelper(root, L, R):
        #     if root == None:
        #         return root
        #     if root.val < L:
        #         return trimBSTHelper(root.right, L, R)
        #     elif root.val > R:
        #         return trimBSTHelper(root.left, L, R)
        #     else:
        #         root.left = trimBSTHelper(root.left, L, root.val)
        #         root.right = trimBSTHelper(root.right, root.val, R)
        #         return root

        def trimBSTHelper(root, L, R):
            if root == None:
                return root
            if root.val < L:
                return trimBSTHelper(root.right, L, R)
            elif root.val > R:
                return trimBSTHelper(root.left, L, R)
            else:
                root.left = trimBSTHelper(root.left, L, root.val)
                root.right = trimBSTHelper(root.right, root.val, R)
                return root

        return trimBSTHelper(root, L, R)

File: Cpp/LC669_Trim_BST.cpp
#include <vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution
{
public:
    // Runtime: 24 ms, faster than 37.12% of C++ online submissions for Trim a Binary Search Tree.
    // Memory Usage: 21.3 MB, less than 100.00% of C++ online submissions for Trim a Binary Search Tree.
    TreeNode *trimBST(TreeNode *root, int L, int R)
    {
        if (!root)
            return nullptr;
        if (root->val < L)
            return trimBST(root->right, L, R);
        else if (root->val > R)
            return trimBST(root->left, L, R);

        root->left = trimBST(root->left, L, R);
        root->right = trimBST(root->right, L, R);
        return root;
    }
};